bin/test_exercise.sh

```diff
@@ -1,112 +1,81 @@
 #!/usr/bin/env bash
 set -eo pipefail
 
-# Test an exercise
+cd "$(git rev-parse --show-toplevel)"
 
-# which exercise are we testing right now?
-# if we were passed an argument, that should be the
-# exercise directory. Otherwise, assume we're in
-# it currently.
-if [ $# -ge 1 ]; then
-    exercise=$1
-    # if this script is called with arguments, it will pass through
-    # any beyond the first to cargo. Note that we can only get a
-    # free default argument if no arguments at all were passed,
-    # so if you are in the exercise directory and want to pass any
-    # arguments to cargo, you need to include the local path first.
-    # I.e. to test in release mode:
-    # $ test_exercise.sh . --release
-    shift 1
-else
-    exercise='.'
+if [ $# -eq 0 ]; then
+    echo "Usage: <exercise-slug> [cargo args]"
+    exit 1
+fi
+
+slug="$1"
+cargo_args="$2"
+
+# determine the exercise path from the slug
+for p in exercises/{practice,concept}/* ; do
+    if [[ "$p" =~ $slug ]]; then
+        exercise_path=$p
+        break
+    fi
+done
+if [ -z "$exercise_path" ]; then
+    echo "Could not find exercise path for $slug"
+    exit 1
 fi
 
 # what cargo command will we use?
 if [ -n "$BENCHMARK" ]; then
-    command="bench"
+    command="+nightly bench"
+    if [ ! -e "$exercise_path/benches" ]; then
+        # no benches, nothing to do
+        exit
+    fi
 else
     command="test"
 fi
 
-declare -a preserve_files=("src/lib.rs" "Cargo.toml" "Cargo.lock")
-declare -a preserve_dirs=("tests")
-
-# reset instructions
-reset () {
-    for file in "${preserve_files[@]}"; do
-        if [ -f "$exercise/$file.orig" ]; then
-            mv -f "$exercise/$file.orig" "$exercise/$file"
-        fi
-    done
-    for dir in "${preserve_dirs[@]}"; do
-        if [ -d "$exercise/$dir.orig" ]; then
-            rm -rf "${exercise:?}/$dir"
-            mv "$exercise/$dir.orig" "$exercise/$dir"
-        fi
-    done
-}
-
-# cause the reset to execute when the script exits normally or is killed
-trap reset EXIT INT TERM
+# setup temporary directory for the exercise
+tmp_dir=$(mktemp -d)
+trap 'rm -rf $tmp_dir' EXIT INT TERM
 
-# preserve the files and directories we care about
-for file in "${preserve_files[@]}"; do
-    if [ -f "$exercise/$file" ]; then
-        cp "$exercise/$file" "$exercise/$file.orig"
-    fi
-done
-for dir in "${preserve_dirs[@]}"; do
-    if [ -d "$exercise/$dir" ]; then
-        cp -r "$exercise/$dir" "$exercise/$dir.orig"
+# copy the exercise to the temporary directory
+contents_to_copy=(
+    "src"
+    "tests"
+    "benches"
+    "Cargo.toml"
+)
+for c in "${contents_to_copy[@]}"; do
+    if [ ! -e "$exercise_path/$c" ]; then
+        continue
     fi
+    cp -r "$exercise_path/$c" "$tmp_dir"
 done
```

Comment:
We shouldn't copy over the entire directory, there may be a large `target` folder.